### PR TITLE
Fix Proposer Bug in Prysm

### LINKED ISSUE #5885: Proposer Index Is Incorrect during Deep Re-orgs
# 🐞 Bug Report

### Description

In schlesi, there has been a current long running fork with more than 80 epochs since finality. This had led to clients being on different side of the fork. As a consequence Prysm has had to process blocks from the other side of the fork. However due to the nature of how long the chain had been forked, each block necessitated a recalculation of a base state from more than 70 epochs back. This has lead us being unable to process the block and the state. The error is stated below.

### Has this worked before in a previous version?

N.A

## 🔬 Minimal Reproduction

Running the slasher will reproduce the error as it tries to retrieve all blocks from genesis and
verify them.

## 🔥 Error

```
time="2020-05-17 00:51:25" level=error msg="Failed to handle p2p pubsub" error="could not process block: could not execute state transition: could not process block in slot 141257: could not process block header: proposer index: 317 is different than calculated: 150" prefix=sync topic="/eth2/9925efd6/beacon_block/ssz_snappy"
```

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Bug Fix 

**What does this PR do? Why is it needed?**

When determining the proposer's index, we simply set the slot in the state and then calculated it. 
However this is incorrect as the state's seed could change during long periods without proposing
a block. This would then lead to an incorrect proposer index being determined.

**Which issues(s) does this PR fix?**

Related to #5885  

**Other notes for review**

### CODE REVIEW

**Comment by terencechain** [comment]
Fixed for @nisdas 

I refactored `packAttestation` to be more efficient. This reduced head state retrieval from 2 to 1

File: beacon-chain/rpc/validator/proposer.go [hunk 1]
```diff
@@ -61,9 +61,17 @@ func (vs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not get ETH1 deposits: %v", err)
 	}
+	head, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get head state %v", err)
+	}
+	head, err = state.ProcessSlots(context.Background(), head, req.Slot)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/proposer.go:68]:
```suggestion
	head, err = state.ProcessSlots(ctx, head, req.Slot)
```

**Comment by prestonvanloon** [comment]
Can someone help me understand the severity of this issue? Do we need to cut a new release ASAP?

**Comment by terencechain** [comment]
> Can someone help me understand the severity of this issue? Do we need to cut a new release ASAP?

This is a consensus breaking bug. @nisdas found this while testing v0.12 with Teku on an interop test chain. This and https://github.com/prysmaticlabs/prysm/pull/6214 both need to go in before genesis. I'd say a release when we know the genesis time is good timing

**Comment by prestonvanloon** [comment]
OK great. It will be nice to get #6221 in as well since the high ETH1 query rates have been bothering users.

**Pull request merged**